miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+    assert mstruct.num == 3
+    mstruct.other = addr2
+    mstruct.deref_other = OtherStruct(jitter.vm, addr)
+
+The `addr` argument can be omited if an allocator is set, in which case the
+structure will be automatically allocated in memory:
+
+    my_heap = miasm2.os_dep.common.heap()
+    set_allocator(my_heap)
+
+Note that some structures (e.g. MemStr or MemArray) do not have a static size
+and cannot be allocated automatically.
+
+
+As you saw previously, to use this module, you just have to inherit from
+MemStruct and define a list of (<field_name>, <field_definition>). Availabe
+MemField classes are:
+
+    - Num: for number (float or int) handling
+    - Struct: abstraction over a simple struct pack/unpack
+    - Ptr: a pointer to another MemStruct instance
+    - Inline: include another MemStruct as a field (equivalent to having a
+      struct field into another struct in C)
+    - Array: a fixed size array of MemFields (points)
+    - Union: similar to `union` in C, list of MemFields at the same offset in a
+      structure; the union has the size of the biggest MemField
+    - BitField: similar to C bitfields, a list of
+      [(<field_name), (number_of_bits)]; creates fields that correspond to
+      certain bits of the field
+
+A MemField always has a fixed size in memory.
+
+
+Some special memory structures are already implemented; they all are subclasses
+of MemStruct with a custom implementation:
+
+    - MemSelf: this class is just a special marker to reference a MemStruct
+      subclass inside itself. Works with Ptr and Array (e.g. Ptr(_, MemSelf)
+      for a pointer the same type as the class who uses this kind of field)
+    - MemVoid: empty MemStruct, placeholder to be casted to an implemented
+      MemStruct subclass
+    - MemStr: represents a string in memory; the encoding can be passed to the
+      constructor (null terminated ascii/ansi or null terminated utf16)
+    - MemArray: an unsized array of MemField; unsized here means that there is
+      no defined sized for this array, equivalent to a int* or char*-style table
+      in C. It cannot be allocated automatically, since it has no known size
+    - MemSizedArray: a sized MemArray, can be automatically allocated in memory
+      and allows more operations than MemArray
+    - mem: a function that dynamically generates a MemStruct subclass from a
+      MemField. This class has only one field named "value".
+
+A MemStruct do not always have a static size (cls.sizeof()) nor a dynamic size
+(self.get_size()).
+"""
+
+import logging
+import struct
+
+log = logging.getLogger(__name__)
+console_handler = logging.StreamHandler()
+console_handler.setFormatter(logging.Formatter("%(levelname)-5s: %(message)s"))
+log.addHandler(console_handler)
+log.setLevel(logging.WARN)
+
+# ALLOCATOR is a function(vm, size) -> allocated_address
+ALLOCATOR = None
+
+# Cache for dynamically generated MemStructs
+DYN_MEM_STRUCT_CACHE = {}
+
+def set_allocator(alloc_func):
```

Comment:
It seems that you only use set_allocator externally, but this method doesn't provide any "improvement" (as check) compared to a `mem.ALLOCATOR = func`. Or you are keeping it for the associated docstring?